modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxUnlockManager.java

```diff
@@ -0,0 +1,321 @@
+
+    private static final int AWAIT_PRIMARY_REPLICA_TIMEOUT = 10;
+
+    private static final int ATTEMPTS_TO_CLEANUP_REPLICA = 5;
+
+    /** Cluster service. */
+    private final ClusterService clusterService;
+    /** Transaction manager. */
+    private final TxManager txManager;
+    /** Lock manager. */
+    private final LockManager lockManager;
+    /**
+     * Placement driver.
+     */
+    private final PlacementDriver placementDriver;
+    /** Hybrid clock. */
+    private final HybridClock hybridClock;
+
+    /**
+     * The constructor.
+     *
+     * @param clusterService Cluster service.
+     * @param txManager Transaction manager.
+     * @param lockManager Lock manager.
+     * @param placementDriver Placement driver.
+     * @param clock A hybrid logical clock.
+     */
+    public TxUnlockManager(
+            ClusterService clusterService,
+            TxManager txManager,
+            LockManager lockManager,
+            PlacementDriver placementDriver,
+            HybridClock clock
+    ) {
+        this.clusterService = clusterService;
+        this.txManager = txManager;
+        this.lockManager = lockManager;
+        this.placementDriver = placementDriver;
+        this.hybridClock = clock;
+    }
+
+    @Override
+    public void start() {
+        clusterService.messagingService().addMessageHandler(TxMessageGroup.class, (msg, sender, correlationId) -> {
+            if (msg instanceof LockReleaseMessage) {
+                processLockRelease((LockReleaseMessage) msg, sender, correlationId);
+            }
+        });
+    }
+
+    /**
+     * Sends unlock request to the primary nodes of each one of {@code partitions}.
+     *
+     * @param partitions Enlisted partition groups.
+     * @param commit {@code true} if a commit requested.
+     * @param commitTimestamp Commit timestamp ({@code null} if it's an abort).
+     * @param txId Transaction id.
+     * @return Completable future of Void.
+     */
+    public CompletableFuture<Void> unlock(
+            Collection<TablePartitionId> partitions,
+            boolean commit,
+            @Nullable HybridTimestamp commitTimestamp,
+            UUID txId
+    ) {
+        Map<TablePartitionId, CompletableFuture<ReplicaMeta>> primaryReplicaFutures = new HashMap<>();
+
+        // Please note that we are using `get primary replica` instead of `await primary replica`.
+        // This method is faster, yet we still have the correctness:
+        // If the primary replica has not changed, get will return a valid value and we'll send an unlock request to this node.
+        // If the primary replica has expired and get returns null (or a different node), the primary node step down logic
+        // will automatically release the locks on that node. All we need to do is to clean the storage.
+        for (TablePartitionId partitionId : partitions) {
+            primaryReplicaFutures.put(partitionId, getPrimaryReplica(partitionId));
```

Comment:
To avoid contention on hybrid clock, please call `clock.now()` once before `for` and then reuse this value in each iteration of the cycle.